lib/Dialect/RTL/Ops.cpp

```diff
@@ -23,28 +29,36 @@ static void buildModule(OpBuilder &builder, OperationState &result,
   result.addAttribute(::mlir::SymbolTable::getSymbolAttrName(), name);
 
   SmallVector<Type, 4> argTypes;
-  for (auto elt : ports)
-    argTypes.push_back(elt.type);
+  SmallVector<Type, 4> resultTypes;
+  for (auto elt : ports) {
+    if (elt.direction == PortDirection::OUTPUT)
+      resultTypes.push_back(elt.type);
+    else
+      argTypes.push_back(elt.type);
+  }
 
   // Record the argument and result types as an attribute.
-  auto type = builder.getFunctionType(argTypes, /*resultTypes=*/{});
+  auto type = builder.getFunctionType(argTypes, resultTypes);
   result.addAttribute(getTypeAttrName(), TypeAttr::get(type));
 
   // Record the names of the arguments if present.
   SmallString<8> attrNameBuf;
   SmallString<8> attrDirBuf;
-  for (size_t i = 0, e = ports.size(); i != e; ++i) {
-    if (ports[i].name.getValue().empty())
-      continue;
-
-    auto argAttr =
-        NamedAttribute(builder.getIdentifier("rtl.name"), ports[i].name);
-
-    auto dirAttr = NamedAttribute(builder.getIdentifier("rtl.direction"),
-                                  ports[i].direction);
-
-    result.addAttribute(getArgAttrName(i, attrNameBuf),
-                        builder.getDictionaryAttr({argAttr, dirAttr}));
+  for (const RTLModulePortInfo &port : ports) {
+    SmallVector<NamedAttribute, 2> argAttrs;
+    if (!port.name.getValue().empty())
+      argAttrs.push_back(
+          NamedAttribute(builder.getIdentifier("rtl.name"), port.name));
+
+    if (port.direction == PortDirection::INOUT)
+      argAttrs.push_back(
+          NamedAttribute(builder.getIdentifier("rtl.inout"),
+                         BoolAttr::get(true, builder.getContext())));
```

Comment:
I'd recommend using a UnitAttr attribute for this, so the presence of the attribute indicates "inout", and the absence indicates "in".  This is `builder.getUnitAttr()`